packages/api/src/module/write/shared/application/events-subscription/events-subscription.ts

```diff
@@ -89,43 +100,36 @@ export class EventsSubscription {
    * See handleEvent for more details how handling event working.
    */
   async start(): Promise<void> {
-    const retryConfig: Partial<RetryConfig<void>> = {
-      ...(this.configuration.options?.retry ?? {
-        retries: 'INFINITELY',
-        backoff: 'EXPONENTIAL',
-        delay: 3000,
-      }),
-      logger: (msg) => this.logger.error(msg),
-      // FIXME: after timeout EventsSubscription will stop processing
-      timeout: 2147483647, // max timeout => max 32-bit signed integer
-    };
+    if (this.running) {
+      this.logger.warn(`${this.subscriptionId} is running`);
 
-    this.eventEmitter.onAny(this.eventEmitterListener);
+      return;
+    }
 
-    // FIXME: it will work for 2147483647/(1000*60*60*24) = 24.85 days :D
-    retry(async () => {
-      await this.catchUp().catch((e) => {
-        this.logger.warn(`EventsSubscription ${this.subscriptionId} processing error in CatchUp phase.`, e);
-        throw e;
-      });
-      await this.listen().catch(async (e) => {
-        this.logger.warn(`EventsSubscription ${this.subscriptionId} processing error in listen phase.`, e);
-        throw e;
-      });
-    }, retryConfig).catch((e) =>
-      this.logger.error(
-        `EventsSubscription ${this.subscriptionId} stopped processing of events after ${retryConfig.retries} retries.`,
-        e,
-      ),
-    );
+    this.logger.debug(`${this.subscriptionId} started`);
+    this.queue.clear();
+    this.eventEmitter.onAny(this.eventEmitterListener);
+    this.shutdownToken.reset();
+    this.running = true;
+    this.run();
   }
 
   /**
    * Stops listening for new events.
    */
   async stop(): Promise<void> {
+    if (!this.running) {
+      this.logger.warn(`${this.subscriptionId} finished`);
+
+      return;
+    }
+
+    this.logger.debug(`${this.subscriptionId} shutdown has been requested`);
     this.eventEmitter.offAny(this.eventEmitterListener);
-    this.queue.clear();
+    this.queue.stop();
+    this.running = false;
+    await this.shutdownToken.wait(60 * 1000);
```

Comment:
What is the purpose of that wait?

Reply:
It's used to wait for current event processing to finish when the stop is requested. Without it, when the stop is requested, then the flow of the currently processing event can be unnecessarily broken. What's more, I think that some unresolved eventHandlers from one test can affect other tests. In that case ShutdownToken could improve stability of tests.

TL;DR
To gracefully shutdown background task ([this.run()](https://github.com/CodersCrew/coderscamp/pull/373/files#diff-549bcd44f6766e2dfd8c0ab2648924875a3972b0a52282e97796215f4120324eR114)).